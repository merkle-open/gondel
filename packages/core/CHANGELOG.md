# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.2.8](https://github.com/merkle-open/gondel/compare/v1.2.7...v1.2.8) (2023-05-19)

### Bug Fixes

- bump deps ([7ae645b](https://github.com/merkle-open/gondel/commit/7ae645ba5d64cf5274fe325c13414cb676b9a615))
- bump deps and default node version ([b06c2d0](https://github.com/merkle-open/gondel/commit/b06c2d0af36b69d0c12fb2263019f08f86b795fa))
- bump deps and default node version ([9c1aeec](https://github.com/merkle-open/gondel/commit/9c1aeec8d3fb03765ea74bb11c29d31205e742bb))
- bump dev deps and add new build files ([98f7ed2](https://github.com/merkle-open/gondel/commit/98f7ed284ab15bd4fc7c2f004c520808f61c6559))
- bump typescript ([f5bbca8](https://github.com/merkle-open/gondel/commit/f5bbca8da39875f21d7bfe6dc1c3076a1e607e15))

## [1.2.7](https://github.com/merkle-open/gondel/compare/v1.2.6...v1.2.7) (2021-05-06)

### Bug Fixes

- property 'align' is missing in type 'HTMLElement' ([c1f460c](https://github.com/merkle-open/gondel/commit/c1f460cbee0e2334c7b52f9e1e43816b27867c28))

## [1.2.6](https://github.com/merkle-open/gondel/compare/v1.2.5...v1.2.6) (2020-11-16)

### Bug Fixes

- improve handling of components with long names ([df63c95](https://github.com/merkle-open/gondel/commit/df63c958f79cfdc5a42b4f710cab3216a35e1529))

## [1.2.5](https://github.com/merkle-open/gondel/compare/v1.2.4...v1.2.5) (2020-11-16)

### Bug Fixes

- allow to stop components in namespaces with long names ([20af9c0](https://github.com/merkle-open/gondel/commit/20af9c09efc4992d48cbc7c300babcffc1db3382))

## [1.2.4](https://github.com/merkle-open/gondel/compare/v1.2.3...v1.2.4) (2020-09-21)

**Note:** Version bump only for package @gondel/core

# [1.2.0](https://github.com/merkle-open/gondel/compare/v1.1.2...v1.2.0) (2020-02-19)

**Note:** Version bump only for package @gondel/core

# [1.1.0](https://github.com/merkle-open/gondel/compare/v1.0.0...v1.1.0) (2020-01-16)

**Note:** Version bump only for package @gondel/core

# [1.0.0](https://github.com/merkle-open/gondel/compare/v0.1.0...v1.0.0) (2019-11-28)

### Bug Fixes

- **core:** prevent plugins from being registered twice when being in different js files ([66aedec](https://github.com/merkle-open/gondel/commit/66aedec)), closes [#48](https://github.com/merkle-open/gondel/issues/48)

### BREAKING CHANGES

- **core:** enhances structure of pluginEvents object to track if a plugin (in another js file) has already been initialized

# [0.1.0](https://github.com/merkle-open/gondel/compare/v0.0.8...v0.1.0) (2019-04-08)

**Note:** Version bump only for package @gondel/core

## [0.0.8](https://github.com/merkle-open/gondel/compare/v0.0.7...v0.0.8) (2018-11-19)

**Note:** Version bump only for package @gondel/core

## [0.0.7](https://github.com/merkle-open/gondel/compare/v0.0.6...v0.0.7) (2018-11-05)

### Bug Fixes

- **core:** Overwrite currentTarget using a getter ([41a859f](https://github.com/merkle-open/gondel/commit/41a859f)), closes [#11](https://github.com/merkle-open/gondel/issues/11)
- **core:** Prevent errors during the event handling ([a560c64](https://github.com/merkle-open/gondel/commit/a560c64))

<a name="0.0.6"></a>

## [0.0.6](https://github.com/merkle-open/gondel/compare/v0.0.5...v0.0.6) (2018-09-25)

### Features

- **core:** Allow to type \_ctx ([0ee58fe](https://github.com/merkle-open/gondel/commit/0ee58fe))

### BREAKING CHANGES

- **core:** GondelBaseComponent constructor requires a ctx and a componentName argument

<a name="0.0.5"></a>

## [0.0.5](https://github.com/merkle-open/gondel/compare/v0.0.4...v0.0.5) (2018-09-19)

### Bug Fixes

- **core:** Allow to stop any node with stopComponents ([4586f89](https://github.com/merkle-open/gondel/commit/4586f89))

<a name="0.0.4"></a>

## [0.0.4](https://github.com/merkle-open/gondel/compare/v0.0.1...v0.0.4) (2018-09-17)

### Bug Fixes

- **@gondel/core:** Rethrow caught errors from promises for better dev tool output ([6de6c5f](https://github.com/merkle-open/gondel/commit/6de6c5f))
- **core:** Allow to overwrite the stop method ([83946fd](https://github.com/merkle-open/gondel/commit/83946fd))
- **core:** Fix Element is undefined in jest ([ba71f7c](https://github.com/merkle-open/gondel/commit/ba71f7c))

### Features

- Upgrade dependencies ([#13](https://github.com/merkle-open/gondel/issues/13)) ([228c287](https://github.com/merkle-open/gondel/commit/228c287))
- **core:** Start all components on dom ready ([12f4b64](https://github.com/merkle-open/gondel/commit/12f4b64))
- **dom-utils:** add the possibility of generic params for dom utils ([63d3785](https://github.com/merkle-open/gondel/commit/63d3785))
- **dom-utils:** refactor getComponentByDomNode for strict typings ([deaf717](https://github.com/merkle-open/gondel/commit/deaf717))
- **events:** Set event.currentTarget to the selected element of the listener ([03a0152](https://github.com/merkle-open/gondel/commit/03a0152))
- **plugin-media-query:** create media-queries gondel plugin ([#2](https://github.com/merkle-open/gondel/issues/2)) ([5c36e04](https://github.com/merkle-open/gondel/commit/5c36e04))

### BREAKING CHANGES

- **events:** Set event.currentTarget to the selected element of the listener
- **dom-utils:** getComponentByDomNode will now throw an error if no component can be found
