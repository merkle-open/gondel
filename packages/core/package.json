{
  "name": "@gondel/core",
  "version": "1.2.7",
  "description": "The base core of the gondel library",
  "bugs": "https://github.com/merkle-open/gondel/issues",
  "homepage": "https://github.com/merkle-open/gondel",
  "repository": "https://github.com/merkle-open/gondel.git",
  "main": "./dist/gondel.es5.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "license": "MIT",
  "scripts": {
    "build": "npm-run-all build:*",
    "build:clean": "rimraf dist",
    "build:types": "tsc --project ./tsconfig.json --declaration --esModuleInterop --target es5 --declarationDir dist/ --sourceMap --outDir dist/",
    "build:rollup": "rollup -c rollup.config.js",
    "build:min": "rollup -c rollup.config.min.js",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "files": [
    "dist/"
  ],
  "devDependencies": {
    "@types/jest": "26.0.23",
    "jest": "26.6.3",
    "npm-run-all": "4.1.5",
    "rimraf": "3.0.2",
    "rollup": "2.58.0",
    "rollup-plugin-uglify": "6.0.4",
    "ts-jest": "26.5.6",
    "tslib": "2.3.1",
    "typescript": "4.4.4"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": true
      }
    },
    "transform": {
      "^.+\\.ts?$": "ts-jest"
    },
    "testRegex": "\\.test\\.ts$",
    "testEnvironment": "jsdom",
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**/*.{ts,tsx,js,jsx}"
    ],
    "coverageDirectory": "coverage",
    "moduleFileExtensions": [
      "ts",
      "js",
      "json",
      "node"
    ],
    "preset": "ts-jest"
  }
}
