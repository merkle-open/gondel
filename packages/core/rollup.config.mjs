export default {
  input: 'dist/index.js',
  output: {
    format: 'umd',
    name: 'gondel',
    file: 'dist/gondel.es5.js',
    sourcemap: true,
    banner: '/// <reference path="./index.d.ts" />'
  }
};
