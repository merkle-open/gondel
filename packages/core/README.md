[![NPM version][npm-image]][npm-url]
[![Build Status](https://github.com/merkle-open/gondel/workflows/ci/badge.svg?branch=master)](https://github.com/merkle-open/gondel/actions)
[![Size][size-image]][size-url]
[![License][license-image]][license-url] 
[![Commitizen friendly][commitizen-image]][commitizen-url] 
[![Prettier](https://img.shields.io/badge/Code%20Style-Prettier-green.svg)](https://github.com/prettier/prettier)

# 🚡 Gondel

[Gondel](https://en.wikipedia.org/wiki/Gondola) is a tiny (3kb) non-intrusive library to help you modularize your code.  
It does **not** ship with a rendering engine to be a perfect fit for most client side rendering engines (e.g. React or Angular) and server side rendering engines (e.g. Java or PHP)

## Installation

```bash
npm i @gondel/core
```

## Hello World

This button will listen to all `click events` events coming from all elements with `data-g-name="Button"` and will
show an alert message.

HTML

```html
 <button data-g-name="Button">Click me</button>

 <button data-g-name="Button">Or click me</button>
```

JS

```js
import {Component, EventListener, GondelBaseComponent} from '@gondel/core';

// The @Component decorator will connect the class with `data-g-name="Button"` elements.
@Component('Button')
export class Button extends GondelBaseComponent {
  @EventListener('click') 
  _handleChange(event) {
    alert('Hello World')
  }
}
```

[Hello World Example](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/hello-world)

## Module format 

Gondel follows the [rollup recommendations](https://github.com/rollup/rollup/wiki/pkg.module) which includes on the one hand ESM for bundle size optimisations and on the other hand a UMD version to be compatible with every former javascript bundling/concatenation strategy.

Gondel is fully typed and exports optional typescript declaration files for typescript projects.

## Plugins

- [Data Plugin](https://github.com/merkle-open/gondel/tree/master/packages/plugins/data) - Provide auto binding of data attributes - [Demo](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/plugin-data)
- [Hot Plugin](https://github.com/merkle-open/gondel/tree/master/packages/plugins/hot) - Adds support for hot-module-reloading (hmr) for Gondel components. - [Demo](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/typescript)
- [jQuery Plugin](https://github.com/merkle-open/gondel/tree/master/packages/plugins/jquery) - Adds support for easy access to the current ctx as jQuery collection.
- [Media Queries Plugin](https://github.com/merkle-open/gondel/tree/master/packages/plugins/media-queries) - Provide a custom gondel event which will fire once a given media query is met - [Demo](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/plugin-media-query)
- [React Plugin](https://github.com/merkle-open/gondel/tree/master/packages/plugins/resize) - Adds support to bootstrap React widgets and apps using Gondel and vice versa - [Demo](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/gondel-react)
- [Resize Plugin](https://github.com/merkle-open/gondel/tree/master/packages/plugins/resize) - Provides an event when the window or the component resized - [Demo](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/plugin-resize)

## Playground

https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/hello-world

## Examples

+ [Hello World](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/hello-world)
+ [Chunk Splitting](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/lazy-load)
+ [Star Rating](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/five-star)
+ [Gondel with Typescript](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/typescript)
+ [Using Gondel from React](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/react-gondel)
+ [Using React from Gondel](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/gondel-react)
+ [Communication getComponentByDomNode](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/get-component-by-dom-node)
+ [Communication findComponents](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/find-components)
+ [Communication triggerPublicEvent](https://codesandbox.io/s/github/merkle-open/gondel/tree/master/examples/trigger-public-event)

## Contributing to Gondel

All contributions are welcome: use-cases, documentation, code, patches, bug reports, feature requests, etc.  
The following commands will get you started to work locally:

```
npm install
npm run build
```

Running tests:

```
npm run test:watch
```

Thanks to all who have contributed ([emoji key](https://allcontributors.org/docs/en/emoji-key)) so far:

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
| [<img src="https://avatars2.githubusercontent.com/u/4113649?v=4" width="100px;"/><br /><sub><b>Jan Nicklas</b></sub>](https://twitter.com/jantimon)<br />[💻](https://github.com/merkle-open/gondel/commits?author=jantimon "Code") [📖](https://github.com/merkle-open/gondel/commits?author=jantimon "Documentation") [🐛](https://github.com/merkle-open/gondel/issues?q=author%3Ajantimon "Bug reports") [💡](#example-jantimon "Examples") [🚇](#infra-jantimon "Infrastructure (Hosting, Build-Tools, etc)") [🔌](#plugin-jantimon "Plugin/utility libraries") [⚠️](https://github.com/merkle-open/gondel/commits?author=jantimon "Tests") [👀](#review-jantimon "Reviewed Pull Requests") | [<img src="https://avatars1.githubusercontent.com/u/149406?v=4" width="100px;"/><br /><sub><b>Ernst Ammann</b></sub>](https://github.com/ernscht)<br />[💻](https://github.com/merkle-open/gondel/commits?author=ernscht "Code") [📖](https://github.com/merkle-open/gondel/commits?author=ernscht "Documentation") [🚇](#infra-ernscht "Infrastructure (Hosting, Build-Tools, etc)") [📦](#platform-ernscht "Packaging/porting to new platform") [👀](#review-ernscht "Reviewed Pull Requests") | [<img src="https://avatars1.githubusercontent.com/u/9339019?v=4" width="100px;"/><br /><sub><b>Dušan Perković</b></sub>](https://github.com/noblica)<br />[💻](https://github.com/merkle-open/gondel/commits?author=noblica "Code") [📖](https://github.com/merkle-open/gondel/commits?author=noblica "Documentation") [🔌](#plugin-noblica "Plugin/utility libraries") [🤔](#ideas-noblica "Ideas, Planning, & Feedback") [⚠️](https://github.com/merkle-open/gondel/commits?author=noblica "Tests") | [<img src="https://avatars1.githubusercontent.com/u/4563751?v=4" width="100px;"/><br /><sub><b>Jan R. Biasi</b></sub>](https://aviormusic.com)<br />[💻](https://github.com/merkle-open/gondel/commits?author=janbiasi "Code") [📖](https://github.com/merkle-open/gondel/commits?author=janbiasi "Documentation") [🤔](#ideas-janbiasi "Ideas, Planning, & Feedback") [⚠️](https://github.com/merkle-open/gondel/commits?author=janbiasi "Tests") [👀](#review-janbiasi "Reviewed Pull Requests") | [<img src="https://avatars1.githubusercontent.com/u/3381820?v=4" width="100px;"/><br /><sub><b>Jan Widmer</b></sub>](https://github.com/janwidmer)<br />[💻](https://github.com/merkle-open/gondel/commits?author=janwidmer "Code") [📖](https://github.com/merkle-open/gondel/commits?author=janwidmer "Documentation") [💡](#example-janwidmer "Examples") [🤔](#ideas-janwidmer "Ideas, Planning, & Feedback") [🔌](#plugin-janwidmer "Plugin/utility libraries") | [<img src="https://avatars1.githubusercontent.com/u/3457712?v=4" width="100px;"/><br /><sub><b>Claudio Bianucci</b></sub>](https://github.com/chezdev)<br />[💻](https://github.com/merkle-open/gondel/commits?author=chezdev "Code") [🤔](#ideas-chezdev "Ideas, Planning, & Feedback") |
| :---: | :---: | :---: | :---: | :---: | :---: |
<!-- ALL-CONTRIBUTORS-LIST:END -->

### License

[MIT license](../../LICENSE)

[npm-image]: https://badge.fury.io/js/%40gondel%2Fcore.svg
[npm-url]: https://npmjs.org/package/@gondel/core
[license-image]: https://img.shields.io/badge/license-MIT-green.svg
[license-url]: http://opensource.org/licenses/MIT
[commitizen-image]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
[commitizen-url]: http://commitizen.github.io/cz-cli/
[size-image]: http://img.badgesize.io/merkle-open/gondel/master/packages/core/dist/gondel.es5.min.js.svg?compression=gzip&label=gzip%20size
[size-url]: https://unpkg.com/@gondel/core/dist/gondel.es5.min.js
