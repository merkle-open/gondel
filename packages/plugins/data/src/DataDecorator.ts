import { GondelComponent } from '@gondel/core';
import { areDataBindingsHookedIntoCore, hookDataDecoratorIntoCore, DataBindingConfig } from './DataPlugin';

export interface ISerializer<T extends any = any> {
	serialize: (value: T) => string;
	deserialize: (value: string) => T;
}

type GondelComponentWithData = GondelComponent & {
	__dataBindings?: Array<DataBindingConfig>;
};

type GondelComponentDecorator<T> = (target: T, propertyKey: string) => void;

/**
 * The @data prop decorator will save the selected value into the given variable at start.
 * Via overloads you're capable to use it in three different ways:
 *  - @data (_)dataMyAttribute => data-my-attribute
 *  - @data('my-attribute') attr => data-my-attribute
 *  - @data('some-json', JSONSerializer) some => data-some-json (serialized via JSON or any other serializer)
 */
export function data<T extends GondelComponentWithData>(target: T, propertyKey: string): void;
export function data<T extends GondelComponentWithData>(
	customAttributeKey: string,
	serializer?: ISerializer
): GondelComponentDecorator<T>;
export function data<T extends GondelComponentWithData>(
	targetOrAttributeKey: T | string,
	propertyKeyOrSerializer: string | ISerializer | undefined
): void | GondelComponentDecorator<T> {
	// First case will be used if we have a custom attribute and a valid serializer (which is typeof ISerializer)
	if (typeof targetOrAttributeKey === 'string' && typeof propertyKeyOrSerializer !== 'string') {
		const customAttributeKey = targetOrAttributeKey;
		const serializer = propertyKeyOrSerializer;

		return function <T extends GondelComponentWithData>(target: T, propertyKey: string): void {
			if (!areDataBindingsHookedIntoCore) {
				// prevent multiple hook listeners
				hookDataDecoratorIntoCore();
			}
			if (!target.__dataBindings) {
				target.__dataBindings = [];
			}
			const attributeKey = `data-${customAttributeKey}`;
			target.__dataBindings.push([propertyKey, attributeKey, serializer]);
		};
	}

	if (typeof targetOrAttributeKey === 'string' || typeof propertyKeyOrSerializer !== 'string') {
		// this case should not occur, the only case could be a respec of the decorators
		throw new Error('Unexpected usage of @data');
	}

	// We only have a simple decorator which will need to autobind values via prop-key
	const target = targetOrAttributeKey;
	const propertyKey = propertyKeyOrSerializer;

	if (!areDataBindingsHookedIntoCore) {
		// prevent multiple hook listeners
		hookDataDecoratorIntoCore();
	}

	if (!target.__dataBindings) {
		target.__dataBindings = [];
	}

	const attributeKey = convertPropertyKeyToDataAttributeKey(propertyKey);
	target.__dataBindings.push([propertyKey, attributeKey, undefined]);
}

/**
 * Will convert any possible property to a valid data attribute
 * @param {string} propertyKey    the prop to convert
 */
function convertPropertyKeyToDataAttributeKey(propertyKey: string): string {
	if (propertyKey.substr(0, 1) === '_') {
		propertyKey = propertyKey.substr(1);
	}

	if (propertyKey.substr(0, 4) !== 'data') {
		throw new Error(
			`${propertyKey}" has an invalid format please use @data dataSomeProp (data-some-prop) for valid bindings.`
		);
	}

	return propertyKey.replace(/([a-zA-Z])(?=[A-Z])/g, '$1-').toLowerCase();
}
