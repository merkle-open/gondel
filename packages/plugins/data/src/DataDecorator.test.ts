import { startComponents, getComponentByDomNode, Component, GondelBaseComponent } from '@gondel/core';
import { data, ISerializer } from './DataDecorator';
import JSONSerializer from './serializer/JSON';

// for custom serializer example
const crypto = require('crypto');
const CRYPTO_ALGORITHM = 'aes-256-ctr';
const CRYPTO_PASSWORD = 'd6F3Efeq';

function createMockElement(namespace: string) {
	const buttonElement = document.createElement('div');
	buttonElement.setAttribute('data-' + namespace + '-name', 'Button');
	buttonElement.setAttribute('data-environment', 'TEST');
	buttonElement.setAttribute('data-text', '2c8bd89c2e314f7f2245e04c'); // encrypted 'Hello World!'
	buttonElement.setAttribute('data-config', JSON.stringify({ hello: 'World!', id: 123 }));
	document.documentElement!.appendChild(buttonElement);
	startComponents(buttonElement, namespace);
	return getComponentByDomNode(buttonElement, namespace)!;
}

describe('@gondel/plugin-data', () => {
	describe('@data', () => {
		it('should automatically convert property key to data value', () => {
			@Component('Button', 'g')
			class Button extends GondelBaseComponent {
				@data
				dataEnvironment: string;
				@data
				dataLanguage: string = 'es';

				getEnv() {
					return this.dataEnvironment;
				}

				getLang() {
					return this.dataLanguage;
				}
			}

			const button = createMockElement('g') as Button;
			expect(button.getLang()).toEqual('es');
			expect(button).toMatchSnapshot();
			expect(button.getEnv()).toBeDefined();
			expect(button.getEnv()).toEqual('TEST');
		});
	});
	it('should use default values from the class if no attr provided', async () => {
		@Component('Button', 'g')
		class Button extends GondelBaseComponent {
			@data
			dataSomeInvalidElement: string = 'initial-value';

			start() {
				expect(this.dataSomeInvalidElement).toEqual('initial-value');
			}
		}

		const button = createMockElement('g') as Button;
		expect(button).toMatchSnapshot();
	});
	it('should allow/auto-convert private-like-values with underscore prefix', () => {
		@Component('Button', 'g')
		class Button extends GondelBaseComponent {
			@data
			_dataEnvironment: string;

			getEnv() {
				return this._dataEnvironment;
			}
		}

		const button = createMockElement('g') as Button;
		expect(button).toMatchSnapshot();
		expect(button.getEnv()).toBeDefined();
		expect(button.getEnv()).toEqual('TEST');
	});
	it('should read values from the data attribute correctly', () => {
		@Component('Button', 'g')
		class Button extends GondelBaseComponent {
			@data('environment')
			private env: string = '';

			getEnv() {
				return this.env;
			}
		}

		const button = createMockElement('g') as Button;
		expect(button).toMatchSnapshot();
		expect(button.getEnv()).toBeDefined();
		expect(button.getEnv()).toEqual('TEST');
	});

	it('should set values inside the data attribute correctly', () => {
		@Component('Button', 'g')
		class Button extends GondelBaseComponent {
			@data('environment')
			private env: string = '';

			setToProd() {
				this.env = `PROD`;
			}

			getEnv() {
				return this.env;
			}
		}

		const button = createMockElement('g') as Button;
		expect(button).toMatchSnapshot();
		expect(button.getEnv()).toBeDefined();
		expect(button.getEnv()).toEqual('TEST');

		button.setToProd();
		expect(button).toMatchSnapshot();
		expect(button.getEnv()).toBeDefined();
		expect(button.getEnv()).toEqual('PROD');
	});

	describe('Serializers', () => {
		it('JSON - should work with JSONs', () => {
			type Config = { hello: string; id: number };
			const initialConfig: Config = { hello: 'World!', id: 123 };
			// const serializedinitialConfig = "{ \"hello\": \"World!\", \"id\": 123 }"; // test-cases

			@Component('Button', 'g')
			class Button extends GondelBaseComponent {
				@data('config', JSONSerializer)
				private config: Config;

				setNewId(id: number) {
					this.config = {
						hello: this.config.hello,
						id,
					};
				}

				getConfig(): Config {
					return this.config;
				}
			}

			const button = createMockElement('g') as Button;
			expect(button).toMatchSnapshot();
			expect(button.getConfig()).toBeDefined();
			expect(button.getConfig()).toEqual(initialConfig);

			button.setNewId(999);
			expect(button).toMatchSnapshot();
			expect(button.getConfig()).toEqual({ ...initialConfig, id: 999 });
		});

		it('Custom - should work with custom serializers (encrypted example)', () => {
			const CustomEncryptionSerializer: ISerializer = {
				serialize(value: string) {
					const cipher = crypto.createCipher(CRYPTO_ALGORITHM, CRYPTO_PASSWORD);
					let crypted = cipher.update(value, 'utf8', 'hex');
					crypted += cipher.final('hex');
					return crypted;
				},
				deserialize(value: string): string {
					const decipher = crypto.createDecipher(CRYPTO_ALGORITHM, CRYPTO_PASSWORD);
					let dec = decipher.update(value, 'hex', 'utf8');
					dec += decipher.final('utf8');
					return dec;
				},
			};

			@Component('Button', 'g')
			class Button extends GondelBaseComponent {
				@data('text', CustomEncryptionSerializer)
				private text: string;

				setText(newText: string) {
					this.text = newText;
				}

				getText(): string {
					return this.text;
				}
			}

			const button = createMockElement('g') as Button;
			expect(button).toMatchSnapshot();
			expect(button.getText()).toBeDefined();
			expect(button.getText()).toEqual('Hello World!');

			button.setText('Hello from the other Side!');
			expect(button).toMatchSnapshot();
			expect(button.getText()).toEqual('Hello from the other Side!');
		});
	});
});
