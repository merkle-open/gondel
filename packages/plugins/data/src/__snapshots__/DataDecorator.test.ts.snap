// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@gondel/plugin-data @data should automatically convert property key to data value 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-language="es"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "dataEnvironment": "TEST",
  "dataLanguage": "es",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data Serializers Custom - should work with custom serializers (encrypted example) 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "stop": [Function],
  "text": "Hello World!",
}
`;

exports[`@gondel/plugin-data Serializers Custom - should work with custom serializers (encrypted example) 2`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e317e623f44a419fae65c94a74781a690d4d46071e8"
  />,
  "_namespace": "g",
  "stop": [Function],
  "text": "Hello from the other Side!",
}
`;

exports[`@gondel/plugin-data Serializers JSON - should work with JSONs 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "config": Object {
    "hello": "World!",
    "id": 123,
  },
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data Serializers JSON - should work with JSONs 2`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":999}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "config": Object {
    "hello": "World!",
    "id": 999,
  },
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data should allow/auto-convert private-like-values with underscore prefix 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_dataEnvironment": "TEST",
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data should read values from the data attribute correctly 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "env": "TEST",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data should set values inside the data attribute correctly 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "env": "TEST",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data should set values inside the data attribute correctly 2`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="PROD"
    data-g-name="Button"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "env": "PROD",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data should use default values from the class if no attr provided 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-config="{\\"hello\\":\\"World!\\",\\"id\\":123}"
    data-environment="TEST"
    data-g-name="Button"
    data-some-invalid-element="initial-value"
    data-text="2c8bd89c2e314f7f2245e04c"
  />,
  "_namespace": "g",
  "dataSomeInvalidElement": "initial-value",
  "stop": [Function],
}
`;
